aurora/core.py

```diff
@@ -555,9 +750,313 @@ def get_par_loss_rate(self, trust_SOL_Ti=False):
+            ELM_energy_delay_param = ELM_period + ELM_energy_delay_param       
+        
+        # Number of time steps during an ELM period
+        time_step = self.namelist['timing']['dt_start'][0]
+        num_time_steps = round(ELM_period/time_step)
+        
+        # build a time-dependent parallel ion flow during an ELM cycle, in a.u.,
+        #    (assuming the ELM crash at t = 0) according to the free streaming model
+        t = np.linspace(time_step,ELM_period,num_time_steps)
+        shape = np.zeros(len(t))
+        for j in range(1,len(t)):
+            shape[j] = (1/((t[j])**2)) * np.exp(-((1/ELM_energy_decay_param)**2)/(2*((t[j])**2)))
+            
+        fraction_to_move = ELM_energy_delay_param/ELM_period
+        indeces_to_move = round(num_time_steps*fraction_to_move)
+        shape_new = np.zeros(len(t))
+        shape_new[indeces_to_move:num_time_steps] = shape[0:num_time_steps-indeces_to_move]
+        shape_new[0:indeces_to_move] = shape[num_time_steps-indeces_to_move:num_time_steps]
+            
+        num_periods = int((self.namelist['timing']['times'][-1]-self.namelist['timing']['times'][0])/ELM_period)
+        
+        shape_new = np.tile(shape_new,num_periods)
+        # make sure that the generated shape has the same length of the time grid
+        if len(shape_new)>len(self.time_grid):      
+            shape_new = shape_new[:-1]
+     
+        return shape_new       
+
+    def ELM_cycle_SOL_mach(self):
+        """
+        Generate a time-dependent shape of the Mach number in the SOL during an ELM
+        cycle, assigning a peak value with a time dependence which follows the parallel
+        ELM flux following the free streaming model.
+        For now this works only if the time grid has constant time steps and for ELMs
+        present for the entire duration of the simulation.
+        """    
+
+        # intra-ELM value of the Mach number
+        M_inter_ELM = self.namelist["SOL_mach"]
+        
+        # requested peak value of the Mach number during an ELM
+        M_peak_intra_ELM = self.namelist["SOL_mach_ELM"]
+        
+        # general time-dependent shape following the ELMS onto the entire time grid
+        shape = self.ELM_cycle_parallel_flux_shape()
+        
+        # normalize the time-dependent shape so that its peaks value equates
+        #   the difference between M_peak_intra_ELM and M_inter_ELM
+        diff = M_peak_intra_ELM - M_inter_ELM
+        max_shape = np.max(shape)
+        ratio = max_shape / diff
+        shape_norm = shape/ratio
+
+        # now add this normalized shape to the inter-ELM value in order to achieve
+        #   a Mach number which flattens on its intra-ELM value during inter-ELM phases
+        #   but peaks at M_peak_intra_ELM in the moment of maximum ELM-carried parallel flux
+        mach = M_inter_ELM + shape_norm
+        
+        return mach   
+    
+    def ELM_cycle_impact_energy_main_wall(self,species):
```

Comment:
@antonellozito could these ELM-specific methods become functions within the `transport_utils.py`, passing only relevant fields as arguments?